### PR TITLE
OEP - implement streaming data-loader for log anomaly detection model training to reduce memory pressure

### PR DESCRIPTION
**Pull request by tybalex**

### CODE REVIEW

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+## Benefits: 
+* resolve the memory pressure issue in the scenario that training data size is large.
+
+## Impact: 
+* This feature will prevent the model training service from running out of memory.
+
+## Implementation details: 
+* the [method that queries training data](https://github.com/rancher/opni-inference-service/blob/main/opnilog-inference-service/opnilog_trainer.py#L46) from Opensearch will return a generator, so it's iterable and it won't keep all the data in memory. It yields each scroll of the search results.
+* To implement a streaming data-loader similar to [`IterableDataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) in this [script](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py). More specifically, an `__iter__()` function will be implemented in this data-loader, which returns an iterator if training dataset.
+* the training method [`train()`](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py#L75) will be slightly adjusted to the new streaming data-loader.
```
Comment (sanjay920) [inline enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:19]: Will you be using Mult-process loading? If so, how many workers and why? Include those details in this section

**Review by sanjay920 — CHANGES_REQUESTED** [review]
Fix inconsistent capitalization in document in addition

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md (hunk 1 above)
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md]: updated.

**Review by AmartC — CHANGES_REQUESTED** [review]
Please address feedback from comments.

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Title: 
+implement streaming data-loader for log anomaly detection model training
+
+## Summary: 
+ AIOps' log anomaly detection model can train model using a streaming data-loader, such that it doesn't have to load all training dataset into memory, instead, it only dumps a subset of training dataset each time and trains with that subset of data. This feature significantly reduces the memory resource pressure in the model training service.
```
[the remark this answers is not in the document]
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md]: fixe.

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+## Use case: 
+When Opni's user launches a workload log anomaly detection training job with huge size of training data (for example, 500Gb logs), the current approach that loads all data into memory won't be feasible. The streaming data-loader can resolve this problem because it only loads a reasonable subset of training dataset each time.
```
[the remark this answers is not in the document]
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md]: updated

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+* Model can be trained correctly from a large training dataset.
+* The memory usage during the model training phase should be within a reasonable range.
+
+
+## Supporting documents: 
+https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset
+
+## Dependencies: 
+
+## Risks and contingencies:
```
[the remark this answers is not in the document]
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:31]: good question. However losing connectivity to Opensearch is not a risk of this feature but rather a risk from using Opensearch.

**Review by dbason — DISMISSED** [review]

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Title: 
+implement streaming data-loader for log anomaly detection model training
```
Comment (AmartC) [inline enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:2]: Minor thing but capitalize the I in Implement

**Review by AmartC — CHANGES_REQUESTED** [review]

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md (hunk 5 above)
Comment (sanjay920) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:2]: @tybalex this hasnt been addressed

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+## Impact: 
+* This feature will prevent the model training service from running out of memory.
+
+## Implementation details: 
+The [method that queries training data](https://github.com/rancher/opni-inference-service/blob/main/opnilog-inference-service/opnilog_trainer.py#L46) from Opensearch will return a generator, so it's iterable and it won't keep all the data in memory. It yields each scroll of the search results.
+* To implement a streaming data-loader similar to [`IterableDataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) in this [script](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py). More specifically, an `__iter__()` function will be implemented in this data-loader, which returns an iterator of training dataset. Multi-processing will be applied to speed up the phase of log masking and weighted random sampling, the number of worker will be the same as the number of CPU cores assigned.
+* The training method [`train()`](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py#L75) will be slightly adjusted to the new streaming data-loader.
```
Comment (sanjay920) [inline enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:19]: Expand on this

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+## Benefits: 
+* Resolve the memory pressure issue in the scenario that training data size is large.
+
+## Impact: 
+* This feature will prevent the model training service from running out of memory.
+
+## Implementation details: 
+The [method that queries training data](https://github.com/rancher/opni-inference-service/blob/main/opnilog-inference-service/opnilog_trainer.py#L46) from Opensearch will return a generator, so it's iterable and it won't keep all the data in memory. It yields each scroll of the search results.
+* To implement a streaming data-loader similar to [`IterableDataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) in this [script](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py). More specifically, an `__iter__()` function will be implemented in this data-loader, which returns an iterator of training dataset. Multi-processing will be applied to speed up the phase of log masking and weighted random sampling, the number of worker will be the same as the number of CPU cores assigned.
```
Comment (sanjay920) [inline enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:18]: We need more implementation details here. A lot of details were discussed in calls - pls add here

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+* The training method [`train()`](https://github.com/rancher/opni-inference-service/blob/main/models/opnilog/opnilog_parser.py#L75) will be slightly adjusted to the new streaming data-loader.
+
+## Acceptance criteria: 
+* Model can be trained correctly from a large training dataset.
+* The memory usage during the model training phase should be within a reasonable range.
```
Comment (sanjay920) [inline enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md:23]: Memory usage should be predictable. An addition acceptance point should be `model training doesnt take longer than before`

**Review by sanjay920 — CHANGES_REQUESTED** [review]

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md (hunk 6 above)
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md]: updated.

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md (hunk 7 above)
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md]: added

File: enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md (hunk 8 above)
Comment (tybalex) [reply enhancements/aiops/20230203-introduce-streaming-dataloader-in-log-anomaly-detection-model-training.md]: addressed

**Review by sanjay920 — APPROVED** [review]

**Review by AmartC — APPROVED** [review]

**Pull request merged**